addons/website/static/src/js/editor/snippets.options.js

```diff
@@ -2822,6 +2822,220 @@ options.registry.ReplaceImage = options.Class.extend({
+                    });
+                }
+                const hideFor = target.dataset[`${attribute.selectName}Rule`] === 'hide';
+                if (hideFor) {
+                    hideAttributes.push({ name: attribute.dataName, records: records});
+                } else {
+                    onlyAttributes.push({ name: attribute.dataName, records: records});
+                }
+                /**
+                 * Create a visibilityId based on the options name and their
+                 * values. eg : hide for en_US(id:1) -> lang1h
+                 */
+                visibilityId += attribute.dataName + records.reduce((acc, record) => acc += record.id, '') + (hideFor ? 'h' : 'o');
+            }
+        }
+        /*
+         * Creates CSS selectors based on those attributes, the reducers
+         * combine the attributes' values.
+         * Important to note how the CSS Rule works : we hide the content
+         * that match the selector, even for "Visible to" option.
+         * Therefore we hide when the HTML tag does not contain the values
+         * we're looking for
+         */
+        let selectors = '';
+        for (const attribute of onlyAttributes) {
+            /**
+             * e.g of selector :
+             * htm:not([data-attr-1="valueAttr1"]) \
+             * :not([data-attr-1="valueAttr2"]) [data-visibility-id="ruleId"]
+             */
+            const selector = attribute.records.reduce((acc, record) => {
+                return acc += `:not([${attribute.name}="${record.value}"])`;
+            }, 'html') + ` body:not(.editor_enable) [data-visibility-id="${visibilityId}"]`;
+            selectors += selector + ', ';
+        }
+        for (const attribute of hideAttributes) {
+            /**
+             * e.g of selector :
+             * html[data-attr-1="valueAttr1"] [data-visibility-id="ruleId"],
+             * html[data-attr-1="valueAttr2"] [data-visibility-id="ruleId"]
+             */
+            const selector = attribute.records.reduce((acc, record, i, a) => {
+                acc += `html[${attribute.name}="${record.value}"] body:not(.editor_enable) [data-visibility-id="${visibilityId}"]`;
+                return acc + (i !== a.length - 1 ? ',' : '');
+            }, '');
+            selectors += selector + ', ';
+        }
+        selectors = selectors.slice(0, -2);
+        if (selectors) {
+            this.$target[0].dataset.visibilitySelectors = selectors;
+        }
+
+        this.$target[0].dataset['visibilityId'] = visibilityId;
+        return await this._super(...arguments);
+    },
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * @override
+     */
+    async _computeWidgetState(methodName, params) {
+        if (methodName === 'selectRecord') {
+            // m2m value widget does not use computeWidgetState
```

Comment:
Why is that actually? Something related to @sdegueldre's code? :/ If yes I really should have paid more attention during the review of its task ^^

I really don't like that `startValue` stuff I see in _renderCustomXML, seems really to be the point of the `_computeWidgetState` method. What if I have another option/widget which tweaks this feature (I don't know, something like a button which "clears all visibility dependencies of this snippet") -> I feel that the m2m widget would still show records even though that button was clicked since _computeWidgetState returns nothing?

If it is something related to SAD's code, could you maybe investigate a bit to see if this can be changed?

Reply:
It seems like I had badly interpreted a comment which said that the initial value does not come from the _computeWidgetState (because the widget will fetch records from the DB). If we return early however I believe the value will still come from _computeWidgetState. I have changed my code to use the _computeWidgetState.